smolyay/grid.py

```diff
@@ -0,0 +1,458 @@
+        self._update(self._dimension)
+
+        return SmolyakGrid(self._grid_points_indexes, self._grid_points,
+                           self._grid_points_basis, self._indices)
+
+    def _update(self, dimension):
+        """Update the properties associated with Smolyak sampling method.
+
+        Update the grid points (and their integers), basis functions and
+        indicies.
+
+        Parameters
+        ----------
+        dimension: int
+            Number of independent variables.
+        """
+        self._dimension = dimension
+        grid_points_indexes = None
+        self._indices = []
+        for sum_of_levels in range(self._dimension,
+                                   self._dimension+len(self._levels)):
+            for composition in generate_compositions(
+                    sum_of_levels, self._dimension, include_zero=False):
+                self._indices.append(composition)
+                # indexes start from zero
+                index_composition = numpy.array(composition) - 1
+                # generate all combinations of
+                # the arrays along each dimension
+                level_composition_index = [self._levels[index]
+                                           for index in index_composition]
+                grid_points_indexes_ = (numpy.array(
+                    numpy.meshgrid(*level_composition_index))
+                    .T.reshape(-1, self._dimension))
+                if grid_points_indexes is None:
+                    grid_points_indexes = grid_points_indexes_
+                else:
+                    grid_points_indexes = numpy.concatenate(
+                        (grid_points_indexes,
+                         grid_points_indexes_), axis=0)
+        # make integer grids, grid points and basis functions
+        self._grid_points_indexes = grid_points_indexes.tolist()
+        self._grid_points = numpy.array(self._points
+                                        )[grid_points_indexes].tolist()
+        self._grid_points_basis = numpy.array(self._basis)[
+                                                grid_points_indexes].tolist()
+
+
+class TensorGridGenerator(IndexGridGenerator):
+    """Create grid points and functions for constructing a surrogate.
+
+    Depending on the dimensionality, points and basis functions provided
+    by :class:`BasisFunctionSet`, the :meth:`__call__` makes full tensor
+    grids.
+    :meth:`__call__`  generates the indexes of
+    grid points (integer grids) depending on dimensionality (full tensor grid).
+    Grid points and their functions are made by replacing the indexes with
+    corresponding point.
+    
+    Parameters
+    ----------
+    basis_set: :class:`BasisFunctionSet`
+        points, basis functions and levels.
+    """
+
+    def __init__(self, basis_set):
+        self._basis_set = basis_set
+        self._points = self._basis_set.points
+        self._basis = self._basis_set.basis_functions
```

Comment:
It would be better to just access these from `self._basis_set`, as it is duplicative to copy them.